libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,919 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  implementation of RunCam camera protocols 
+
+  With thanks to betaflight for a great reference
+  implementation. Several of the functions below are based on
+  betaflight equivalent functions
+
+  RunCam protocol specification can be found at https://support.runcam.com/hc/en-us/articles/360014537794-RunCam-Device-Protocol
+ */
+#include "AP_RunCam.h"
+
+#include <AP_Math/AP_Math.h>
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Logger/AP_Logger.h>
+
+const AP_Param::GroupInfo AP_RunCam::var_info[] = {
+    // @Param: FEATURES
+    // @DisplayName: RunCam features available
+    // @Description: The available features of the attached RunCam device. If 0 then the RunCam device will be queried for the features it supports, otherwise this setting is used.
+    // @User: Advanced
+    // @Bitmask: 0:Power Button,1:WiFi Button,2:Change Mode,3:5-Key OSD,4:Settings Access,5:DisplayPort,6:Start Recording,7:Stop Recording
+    AP_GROUPINFO("FEATURES", 1, AP_RunCam, _features, 0),
+
+    // @Param: BT_DELAY
+    // @DisplayName: RunCam boot delay before allowing updates
+    // @Description: Time it takes for the RunCam to become fully ready in ms. If this is too short then commands can get out of sync.
+    // @User: Advanced
+    AP_GROUPINFO("BT_DELAY", 2, AP_RunCam, _boot_delay_ms, 6000),
+
+    // @Param: BTN_DELAY
+    // @DisplayName: RunCam button delay before allowing further button presses
+    // @Description: Time it takes for the a RunCam button press to be actived in ms. If this is too short then commands can get out of sync.
+    // @User: Advanced
+    AP_GROUPINFO("BTN_DELAY", 3, AP_RunCam, _button_delay_ms, 300),
+
+    AP_GROUPEND
+};
+
+#define RUNCAM_DEBUG 0
+#define RUNCAM_INIT_INTERVAL_MS 500
+#define RUNCAM_OSD_TICK_COUNT     5
+
+#if RUNCAM_DEBUG
+#define debug(fmt, args ...) do { hal.console->printf("RunCam: " fmt, ## args); } while (0)
+#else
+#define debug(fmt, args ...)
+#endif
+
+extern const AP_HAL::HAL& hal;
+
+// singleton instance
+AP_RunCam *AP_RunCam::_singleton;
+
+AP_RunCam::Request::Length AP_RunCam::Request::_expected_responses_length[RUNCAM_NUM_EXPECTED_RESPONSES] = {
+    { Command::RCDEVICE_PROTOCOL_COMMAND_GET_DEVICE_INFO, 5 },
+    { Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_PRESS, 2 },
+    { Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_RELEASE, 2 },
+    { Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION, 3 },
+};
+
+// these are correct for the runcam split micro v2.4.4, others may vary
```

Comment:
What happens if the menus are of different lengths?

Is it firmware-version dependant?

Reply:
I don't know but a description of a menu I found for older firmware was different, so probably. Ultimately this should be configurable, but not quite sure how